// @colyseus/redis-driver/build/RoomData.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RoomData_exports = {};
__export(RoomData_exports, {
  RoomData: () => RoomData
});
module.exports = __toCommonJS(RoomData_exports);
var import_core = require("@colyseus/core");
class RoomData {
  clients = 0;
  locked = false;
  private = false;
  maxClients = Infinity;
  metadata;
  name;
  publicAddress;
  processId;
  roomId;
  createdAt;
  unlisted = false;
  #client;
  #removed = false;
  constructor(initialValues, client) {
    this.#client = client;
    this.createdAt = initialValues && initialValues.createdAt ? new Date(initialValues.createdAt) : new Date();
    for (const field in initialValues) {
      if (initialValues.hasOwnProperty(field)) {
        this[field] = initialValues[field];
      }
    }
  }
  toJSON() {
    return {
      clients: this.clients,
      createdAt: this.createdAt,
      maxClients: this.maxClients,
      metadata: this.metadata,
      name: this.name,
      publicAddress: this.publicAddress,
      processId: this.processId,
      roomId: this.roomId
    };
  }
  async save() {
    if (this.#removed) {
      return;
    }
    if (this.roomId) {
      const toJSON = this.toJSON;
      this.toJSON = void 0;
      const roomcache = JSON.stringify(this);
      this.toJSON = toJSON;
      await this.hset("roomcaches", this.roomId, roomcache);
    } else {
      import_core.logger.warn("RedisDriver: can't .save() without a `roomId`");
    }
  }
  updateOne(operations) {
    if (operations.$set) {
      for (const field in operations.$set) {
        if (operations.$set.hasOwnProperty(field)) {
          this[field] = operations.$set[field];
        }
      }
    }
    if (operations.$inc) {
      for (const field in operations.$inc) {
        if (operations.$inc.hasOwnProperty(field)) {
          this[field] += operations.$inc[field];
        }
      }
    }
    return this.save();
  }
  remove() {
    if (this.roomId) {
      this.#removed = true;
      return this.hdel("roomcaches", this.roomId);
    }
  }
  async hset(key, field, value) {
    return await this.#client.hset(key, field, value);
  }
  async hdel(key, field) {
    return await this.#client.hdel(key, field);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RoomData
});
