// @colyseus/core/build/Transport.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Transport_exports = {};
__export(Transport_exports, {
  ClientArray: () => ClientArray,
  ClientState: () => ClientState,
  Transport: () => Transport
});
module.exports = __toCommonJS(Transport_exports);
var import_Utils2 = require("./Utils");
class Transport {
  server;
}
var ClientState = /* @__PURE__ */ ((ClientState2) => {
  ClientState2[ClientState2["JOINING"] = 0] = "JOINING";
  ClientState2[ClientState2["JOINED"] = 1] = "JOINED";
  ClientState2[ClientState2["RECONNECTED"] = 2] = "RECONNECTED";
  ClientState2[ClientState2["LEAVING"] = 3] = "LEAVING";
  return ClientState2;
})(ClientState || {});
class ClientArray extends Array {
  getById(sessionId) {
    return this.find((client) => client.sessionId === sessionId);
  }
  delete(client) {
    return (0, import_Utils2.spliceOne)(this, this.indexOf(client));
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ClientArray,
  ClientState,
  Transport
});
